### ISSUE TITLE
bug/jenkins: spawn/disconnect 3 hangs with shmem config

### ISSUE THREAD

**Issue by hzhou** [pmodels/mpich#4516]
With `ch4:ofi` and without `MPIR_CVAR_ODD_EVEN_CLIQUES=1`:
```
not ok  - ./spawn/disconnect3 3
  ---
  Directory: ./spawn
  File: disconnect3
  Num-procs: 3
  Timeout: 180
  Date: "Fri May  8 12:45:20 2020"
  ...
## Test output (expected 'No Errors'):
## [mpiexec@pmrs-centos64-240-02.cels.anl.gov] APPLICATION TIMED OUT, TIMEOUT = 180s
```

**Comment by hzhou** [comment]
It hangs due to one of the process finalize and exit before an ofi lightweight send data gets flushed out and consequently the receiving process hangs at `MPI_Recv`.

In `MPIDI_OFI_finalize_hook`, we currently use an "allreduce" to flush all send data out. This works when two processes are consistently communicate over netmod or shmmod. Those two processes on a same node never communicate over ofi and don't have send data to flush. However, all dynamic processes communicate over netmod to parents, so in this example, process 1 and 2 communicate over ofi between intercomm, but communicate over shmmod in allreduce intracomm. So the allreduce in finalize doesn't flush the process 2's send data because it's skipping it.

In general, this using "allreduce" to hack flushing send data is a hack and unreliable. For example, once we have multiple VCI, a single allreduce is not going to flush out all vci channels anyway. What we needed is an explicit send flush.

PR to fix this coming.

**Comment by zhenggb72** [comment]
Should we consider a counter-based scheme (or a quiescence detection algorithm) which keeps track of pending sends/receives to wait before starting finalizing?
I feel any reduction collective-based fix may work for one workload but may not work for the other. We may have to flush messages for both netmod and shmmod.

**Comment by hzhou** [comment]
@zhenggb72 `shmmod` doesn't need flush. The problem is not message sent over shmem got lost. It is the message sent over netmod -- the one sent to another process via intercom -- and ofi never flushed its queue since `Allreduce` goes through shmmod and didn't touch netmod message queue.

**Comment by zhenggb72** [comment]
Why shmmod does not need flush? For example, if it is a IPC-based send, the receiver may still getting the data while the sender may end its execution without checking.

**Comment by hzhou** [comment]
With shmem, sender will not exit unless either the entire message is placed into the shared memory or the IPC has received ack from receiver. Either case the message is sent out before exit.

**Comment by zhenggb72** [comment]
I am confused, I see we do have postponed queue for sends and progress_send in posix. And I thought we should also progress IPC as well when it is a non-blocking call (I see we don't, but why?).

**Comment by hzhou** [comment]
> In any case, even just for future-proof, I am not sure it is a good idea to assume shmmod does not need flush.

You are sidetracking. How about let's focus on the issue at hand and not to worry about *future* issues that we doesn't have at the moment?

> I am confused,

Which part are you still confused? Are you still questioning whether the bug in this issue is from shmmod or ofi? Well, if it is in shmmod, the fix in OFI won't fix it, right? I think I proved that by PR #5167.

**Comment by zhenggb72** [comment]
I am not sure what you mean by sidetracking. My comments above combines what I read here and the proposed fix PR (which does the flush send/recv, i.e. only flush netmod). Please read it again in case you missed my edits earlier.

While the old reduction-based implementation may be only doing shmmod "flush" (when on one node) and skipped netmod, the proposed fix only does netmod "flush", and hence my concern above.

**Comment by hzhou** [comment]
> ... and the proposed fix PR (which does the flush send/recv, i.e. only flush netmod).

So you will be pushing that PR?

**Comment by zhenggb72** [comment]
You mean approve?

**Comment by hzhou** [comment]
I thought you are proposing a new solution, maybe I miss-read.
